File: app/models/player.py
from .db import db
from .player_team import player_teams

class Player(db.Model):
    __tablename__ = 'players'

    id = db.Column(db.Integer, primary_key=True)
    league_id = db.Column(db.Integer, db.ForeignKey('leagues.id'), nullable=False)
    player_name = db.Column(db.String(40), nullable=False)
    player_image = db.Column(db.String(500), nullable=True, default='https://esrf.s3.amazonaws.com/Empty-Player-Image.png')
    position = db.Column(db.String(20), nullable=False)
    team = db.Column(db.String(40), nullable=True, default='')
    bio = db.Column(db.String(1000), nullable=True, default='')
    recent_news = db.Column(db.String(1000), nullable=True, default='')
    field_goal_made = db.Column(db.Integer, nullable=True, default=0)
    field_goal_attempted = db.Column(db.Integer, nullable=True, default=0)
    free_throw_made = db.Column(db.Integer, nullable=True, default=0)
    free_throw_attempted = db.Column(db.Integer, nullable=True, default=0)
    three_point_made = db.Column(db.Integer, nullable=True, default=0)
    assists = db.Column(db.Integer, nullable=True, default=0)
    rebounds = db.Column(db.Integer, nullable=True, default=0)
    steals = db.Column(db.Integer, nullable=True, default=0)
    blocks = db.Column(db.Integer, nullable=True, default=0)
    turnovers = db.Column(db.Integer, nullable=True, default=0)
    points = db.Column(db.Integer, nullable=True, default=0)
    fantasy_total = db.Column(db.Integer, nullable=True, default=0)

    player_league = db.relationship('League', back_populates='league_players')
    teams_with_player = db.relationship('Team',
    secondary=player_teams,
    back_populates='players_on_team'
    )


    def to_dict_no_team(self):
        return {
            'id': self.id,
            'league_id': self.league_id,
            'player_name': self.player_name,
            'player_image': self.player_image,
            'position': self.position,
            'team': self.team,
            'bio': self.bio,
            'recent_news': self.recent_news,
            'field_goal_made': self.field_goal_made,
            'field_goal_attempted': self.field_goal_attempted,
            'free_throw_made': self.free_throw_made,
            'free_throw_attempted': self.free_throw_attempted,
            'three_point_made': self.three_point_made,
            'assists': self.assists,
            'rebounds': self.rebounds,
            'steals': self.steals,
            'blocks': self.blocks,
            'turnovers': self.turnovers,
            'points': self.points,
            'fantasy_total': self.fantasy_total,
        }

    def to_dict(self):
        return {
            'id': self.id,
            'league_id': self.league_id,
            'player_name': self.player_name,
            'player_image': self.player_image,
            'position': self.position,
            'team': self.team,
            'bio': self.bio,
            'recent_news': self.recent_news,
            'field_goal_made': self.field_goal_made,
            'field_goal_attempted': self.field_goal_attempted,
            'free_throw_made': self.free_throw_made,
            'free_throw_attempted': self.free_throw_attempted,
            'three_point_made': self.three_point_made,
            'assists': self.assists,
            'rebounds': self.rebounds,
            'steals': self.steals,
            'blocks': self.blocks,
            'turnovers': self.turnovers,
            'points': self.points,
            'fantasy_total': self.fantasy_total,
            'team_list': [team.to_dict_for_player() for team in self.teams_with_player]
        }

File: react-app/src/store/player.js
// Read
const GET_PLAYERS = 'players/GET_PLAYERS';

// Create and Edit
const ADD_PLAYER = 'players/ADD_PLAYER';

// Delete
const DELETE_PLAYER = 'players/DELETE_PLAYER';

// Clear
const CLEAR_PLAYERS = 'players/CLEAR_PLAYERS';


// Thunk Action Creators
const actionGetPlayers = (players) => ({
    type: GET_PLAYERS,
    players
})

export const actionAddPlayer = (player) => ({
    type: ADD_PLAYER,
    player
})

const actionDeletePlayer = (playerId) => ({
    type: DELETE_PLAYER,
    playerId
})

export const clearPlayers = () => ({
    type: CLEAR_PLAYERS
})


// Thunks
export const getAllPlayers = (leagueId) => async (dispatch) => {
    const response = await fetch(`/api/players/league/${leagueId}`);

    if (response.ok) {
        const players = await response.json();
        dispatch(actionGetPlayers(players));
        return players;
    }
}

export const addPlayer = (league_id, player_name, position, team, bio) => async (dispatch) => {
    const response = await fetch('/api/players/new', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            league_id,
            player_name,
            position,
            team,
            bio
        })
    });

    if (response.ok) {
        const player = await response.json();
        dispatch(actionAddPlayer(player));
        return player;
    }
}

export const editPlayerInfo = (playerId, player_name, position, team, bio) => async (dispatch) => {
    const response = await fetch(`/api/players/edit/${playerId}`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            player_name,
            position,
            team,
            bio
        })
    });

    if (response.ok) {
        const player = await response.json();
        dispatch(actionAddPlayer(player));
        return player;
    }
    else {
        const errors = await response.json();
        return errors
    }
}

export const editPlayerStats = (payload) => async (dispatch) => {
    const playerId = payload.playerId;
    const recent_news = payload.recent_news;
    const field_goal_made = payload.field_goal_made;
    const field_goal_attempted = payload.field_goal_attempted;
    const free_throw_made = payload.free_throw_made;
    const free_throw_attempted = payload.free_throw_attempted;
    const three_point_made = payload.three_point_made;
    const assists = payload.assists;
    const rebounds = payload.rebounds;
    const steals = payload.steals;
    const blocks = payload.blocks;
    const turnovers = payload.turnovers;
    const points = payload.points;

    const response = await fetch(`/api/players/edit/${playerId}/stats`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            recent_news,
            field_goal_made,
            field_goal_attempted,
            free_throw_made,
            free_throw_attempted,
            three_point_made,
            assists,
            rebounds,
            steals,
            blocks,
            turnovers,
            points
        })
    });

    if (response.ok) {
        const player = await response.json();
        dispatch(actionAddPlayer(player));
        return player;
    };
}

export const deletePlayer = (formData) => async (dispatch) => {
    const response = await fetch(`/api/players/delete`, {
        method: 'DELETE',
        body: formData
    });

    if (response.ok) {
        const playerId = await response.json();
        dispatch(actionDeletePlayer(playerId));
    } else {
        const errors = await response.json();
        return errors;
    }
}


// Player Reducer
const PlayersReducer = (state = {}, action) => {
    switch (action.type) {
        case GET_PLAYERS:
            const newState1 = {};
            action.players.playerList.forEach(player => {
                newState1[player.id] = player;
            });
            return newState1;
        case ADD_PLAYER:
            const newState2 = { ...state };
            newState2[action.player.id] = action.player;
            return newState2;
        case DELETE_PLAYER:
            const newState3 = { ...state };
            delete newState3[action.playerId];
            return newState3;
        case CLEAR_PLAYERS:
            const clearState = {};
            return clearState;
        default:
            return state;
    }
}

export default PlayersReducer;
